lib/streamlit/connections/snowpark_connection.py

```diff
@@ -0,0 +1,101 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# NOTE: We won't always be able to import from snowflake.snowpark.session so need the
+# `type:ignore` comment below, but that comment will explode if `warn-unused-ignores` is
+# turned on when the package is available. Unfortunately, mypy doesn't provide a good
+# way to configure this at a per-line level :(
+# mypy: no-warn-unused-ignores
+
+import configparser
+import os
+import threading
+from contextlib import contextmanager
+from datetime import timedelta
+from typing import TYPE_CHECKING, Any, Dict, Iterator, Optional, Union, cast
+
+import pandas as pd
+
+from streamlit.connections import BaseConnection
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_data
+
+if TYPE_CHECKING:
+    from snowflake.snowpark.session import Session  # type: ignore
+
+
+_REQUIRED_CONNECTION_PARAMS = {"account", "user"}
+_DEFAULT_CONNECTION_FILE = "~/.snowsql/config"
+
+
+def _load_from_snowsql_config_file() -> Dict[str, Any]:
+    """Loads the dictionary from snowsql config file."""
+    snowsql_config_file = os.path.expanduser(_DEFAULT_CONNECTION_FILE)
+    if not os.path.exists(snowsql_config_file):
+        return {}
+
+    config = configparser.ConfigParser(inline_comment_prefixes="#")
+    config.read(snowsql_config_file)
+
+    conn_params = config["connections"]
+    conn_params = {k.replace("name", ""): v.strip('"') for k, v in conn_params.items()}
+    if "db" in conn_params:
+        conn_params["database"] = conn_params["db"]
+        del conn_params["db"]
+    return conn_params
+
+
+class Snowpark(BaseConnection["Session"]):
+    _default_connection_name = "snowpark"
+
+    def __init__(self, connection_name: str = "default", **kwargs) -> None:
+        self._lock = threading.RLock()
+
+        # Grab the lock before calling BaseConnection.__init__() so that we can guarantee
+        # thread safety when the parent class' constructor initializes our connection.
+        with self._lock:
+            super().__init__(connection_name, **kwargs)
+
+    # TODO(vdonato): Teach the .connect() method how to automagically connect in a SiS
+    # runtime environment.
+    def connect(self, **kwargs) -> "Session":
+        from snowflake.snowpark.session import Session
+
+        conn_params = self.get_secrets().__nested_secrets__
```

Comment:
When I was implementing one yesterday I wanted to merge the secrets with the args and ended up writing a helper function to convert the streamlit secrets format into a dict

I also got tripped up by it looking like a AttrDict but the [merging behavior](https://github.com/bcj/AttrDict#merging) I found online not working.

I think I have a slightly different philosophy from Vincent on merging secrets and args but just to share that datapoint :)

Reply:
This should be made better by #6387. I'm debating whether to call this at the end of `get_secrets` so that the user always gets a regular dict, but I think it'd be nice to be able to interact with secrets subsections similarly to how you can if you fetched them directly from `st.secrets`.

It'll be a pretty trivial change if we decide to just give users a plain dict, so can assume we'll be returning an `AttrDict` for now and we can easily change our  mind later